# Indekseringsoppgaven

Generelle regler for indeksering i Python:
- Når man angir indekser i Python bruker man brakkeparanteser []. 
- Indekser MÅ være integers - altså det kan ikke være noen desimaler, og det holder ikke at desimalen er 0. 
- Python starter indekseringen sin på 0. 
- Den siste indeksen er -1. Den nest siste er -2 etc. 
- Skal man ha alle indekser i en rad eller kolonne bruker man kolon.
- I en matrise med to dimensjoner er den første indeksen alltid radene, mens den andre indeksen er kolonnene.
- Skal man ha tak i mer enn en, men ikke alle, elementene skriver man [a:b]. Dette betyr element a til *men ikke med* b. 
- Hvis man har en matrise og skal ha tak i rad a til men ikke med b, og alle kolonnene skriver man  [a:b,:]. Rader og kolonner skilles altså med komma. 
- Vil man ha fra start frem til (og ikke med) en indeks skriver man [:a]. Vil man ha fra (og med) en indeks til slutten skriver man [a:]. 

Vi bruker datasettet over årlig temperatur i Bergen til å gå gjennom disse punktene. 

In [1]:
import numpy as np

Last inn datasettet "TempBergenYearly.txt". Se på oppgaven "RyddDatasett" for flere detaljer på hvordan filinnlastingen fungerer. 

In [12]:
file='TempBergenYearly.txt'
data = np.genfromtxt('https://raw.githubusercontent.com/irendundas/EkteData/'
                     'main/'+file, dtype=float, delimiter=',',skip_header=1)
print(data.shape)
data

(161, 2)


array([[1861.        , -999.99      ],
       [1862.        ,    7.71666667],
       [1863.        ,    7.75833333],
       [1864.        ,    6.86666667],
       [1865.        ,    7.225     ],
       [1866.        ,    7.41666667],
       [1867.        ,    6.55      ],
       [1868.        ,    7.975     ],
       [1869.        ,    6.45      ],
       [1870.        ,    6.575     ],
       [1871.        , -999.99      ],
       [1872.        , -999.99      ],
       [1873.        ,    7.83333333],
       [1874.        ,    7.175     ],
       [1875.        ,    7.01666667],
       [1876.        ,    6.78333333],
       [1877.        ,    6.55      ],
       [1878.        , -999.99      ],
       [1879.        ,    6.69166667],
       [1880.        ,    7.53333333],
       [1881.        ,    6.36666667],
       [1882.        ,    7.9       ],
       [1883.        ,    7.6       ],
       [1884.        ,    8.20833333],
       [1885.        ,    6.70833333],
       [1886.        ,   

data.shape viser oss at datasettet over temperatur har 161 rader og 2 kolonner. Hvis vi vil ha tak i alle årene og lagre dem i en vektor som vi kaller "tid" skriver vi `tid=data[:,0]`, altså *alle radene* (:), men kun den *første kolonnen* (0)


In [13]:
tid=data[:,0]
tid

array([1861., 1862., 1863., 1864., 1865., 1866., 1867., 1868., 1869.,
       1870., 1871., 1872., 1873., 1874., 1875., 1876., 1877., 1878.,
       1879., 1880., 1881., 1882., 1883., 1884., 1885., 1886., 1887.,
       1888., 1889., 1890., 1891., 1892., 1893., 1894., 1895., 1896.,
       1897., 1898., 1899., 1900., 1901., 1902., 1903., 1904., 1905.,
       1906., 1907., 1908., 1909., 1910., 1911., 1912., 1913., 1914.,
       1915., 1916., 1917., 1918., 1919., 1920., 1921., 1922., 1923.,
       1924., 1925., 1926., 1927., 1928., 1929., 1930., 1931., 1932.,
       1933., 1934., 1935., 1936., 1937., 1938., 1939., 1940., 1941.,
       1942., 1943., 1944., 1945., 1946., 1947., 1948., 1949., 1950.,
       1951., 1952., 1953., 1954., 1955., 1956., 1957., 1958., 1959.,
       1960., 1961., 1962., 1963., 1964., 1965., 1966., 1967., 1968.,
       1969., 1970., 1971., 1972., 1973., 1974., 1975., 1976., 1977.,
       1978., 1979., 1980., 1981., 1982., 1983., 1984., 1985., 1986.,
       1987., 1988.,

Lag en variabel "temp" som inneholder kolonnen med temperatur.

Skriv ut årene og temperatur for alle årene fra året du ble født frem til slutten av datasettet. Da må du finne indeksen til året du ble født - husk at `tid[0]` gir deg startåret. Når du regner deg frem til en indeks-verdi er gjerne resultatet en "float" (tall med desimal). For å konvertere en float til integer kan du skrive int(a). 

Finn de fem siste årene i datasettet.

Finn tidsserien for temperatur på hele 1900-tallet. Skriv også ut årene. 

Iblant trenger man å hente ut data med jevne intervall - f.eks. hvis man har et datasett med daglig oppløsning og bare vil ha tak i onsdagene, eller man bare vil ha tak i bursdagen sin fra en tidsserie med daglig oppløsning som spenner over flere år. Eller kanskje man vil hente ut data fra kl 12 for å sjekke solstyrke midt på dagen... 

Med indeksering bruker man tre verdier: [start:stop:step]. Skal man f.eks. ha annenhver verdi mellom indeks a og b skriver man [a:b:2]. Husk at det fremdeles er a *til og ikke med* b. 

Finn temperaturen for hvert tiende år siden 1870 frem til 2020. Skriv igjen ut årene i tillegg (for å sjekke at indekseringen du har satt stemmer).

Lag en figur som viser hele temperaturtidsserien, og plot hvert tiende år på samme figur. Synes du tidsserien blir godt representert ved å bare se på hvert tiende år? I en seinere oppgave, f.eks. den om for-løkker, ser vi på hvordan man kan representere datasettet ved å ta gjennomsnittet over tiårsperioder. Hvordan tror du denne tidsserien representerer datasettet i forhold til den figuren vi har her?

In [8]:
tid[0]

1861.0

Plukk ut bare april